Address remaining safer CPP warnings in WebKit/Shared

cdumez opened the pull request:
#### 520ff168966e217ac220e6874bfb46087e6d546d
<pre>
Address remaining safer CPP warnings in WebKit/Shared
<a href="https://bugs.webkit.org/show_bug.cgi?id=301163">https://bugs.webkit.org/show_bug.cgi?id=301163</a>

Reviewed by Timothy Hatcher and Darin Adler.

* Source/WTF/wtf/text/WTFString.h:
(WTF::nsStringNilIfEmpty):
(WTF::nsStringNilIfNull):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _accessibilityHitTestResolvingRemoteFrame:callback:]):
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::LinkImageLayout::LinkImageLayout):
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::protectedPlatformPath const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setVideoGravity):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.mm:
(WebCore::convertMediaPlayerToAVLayerVideoGravity):
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::ResourceError):
(WebCore::ResourceError::mapPlatformError):
(WebCore::ResourceError::cfError const):
(WebCore::ResourceError::protectedCFError const):
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::convertWTFStringToNSString):
* Source/WebKit/Platform/ios/PaymentAuthorizationController.mm:
(-[WKPaymentAuthorizationControllerDelegate presentationSceneIdentifierForPaymentAuthorizationController:]):
(-[WKPaymentAuthorizationControllerDelegate presentationSceneBundleIdentifierForPaymentAuthorizationController:]):
* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/NoUnretainedMemberCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(encodeObject):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(-[_WKRemoteObjectInterface initWithProtocol:identifier:]):
(-[_WKRemoteObjectInterface protocol]):
(-[_WKRemoteObjectInterface debugDescription]):
* Source/WebKit/Shared/API/c/cf/WKStringCF.mm:
(WKStringCreateWithCFString):
(wkNSStringClassSingleton): Deleted.
* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCreateWithCFURL):
(wkNSURLClassSingleton): Deleted.
* Source/WebKit/Shared/API/c/mac/WKURLRequestNS.mm:
(WKURLRequestCopyNSURLRequest):
* Source/WebKit/Shared/API/c/mac/WKURLResponseNS.mm:
(WKURLResponseCopyNSURLResponse):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
(WebKit::CoreIPCCNPostalAddress::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPassKit.mm:
(WebKit::CoreIPCPKContact::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.mm:
(WebKit::CoreIPCPersonNameComponents::toID const):
* Source/WebKit/Shared/Cocoa/RevealItem.h:
* Source/WebKit/Shared/Cocoa/RevealItem.mm:
(WebKit::RevealItem::highlightRange const):
(WebKit::RevealItem::protectedItem const):
* Source/WebKit/Shared/Cocoa/WKNSError.mm:
(-[WKNSError _web_createTarget]):
* Source/WebKit/Shared/Cocoa/WKNSURLRequest.mm:
(-[WKNSURLRequest _web_createTarget]):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/Daemon/PCMDaemonEntryPoint.mm:
(WebKit::CompletionHandler&lt;void):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::toCAFilterType):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::launchServicesCheckIn):
(WebKit::webKit2BundleSingleton):
(WebKit::getSandboxProfileOrProfilePath):
(WebKit::populateSandboxInitializationParameters):
(WebKit::webKit2Bundle): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension defaultLocale]):
(-[WKWebExtension displayName]):
(-[WKWebExtension displayShortName]):
(-[WKWebExtension displayVersion]):
(-[WKWebExtension displayDescription]):
(-[WKWebExtension displayActionLabel]):
(-[WKWebExtension version]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getTextFragmentMatchWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _applicationNameForDesktopUserAgent]):
(-[WKWebViewConfiguration applicationNameForUserAgent]):
(-[WKWebViewConfiguration _groupIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo qrCodePayloadString]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _defaultSTSLabel]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(+[WKAutocorrectionContext autocorrectionContextWithWebContext:]):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem URLString]):
(-[WebHistoryItem originalURLString]):
(-[WebHistoryItem title]):
(-[WebHistoryItem alternateTitle]):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(-[WebHTMLRepresentation title]):
* Source/WebKitLegacy/mac/WebView/WebScriptDebugDelegate.mm:
(-[WebScriptCallFrame functionName]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _mainFrameOverrideEncoding]):

Canonical link: <a href="https://commits.webkit.org/301920@main">https://commits.webkit.org/301920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f2a015390d026f738b6e2da7b8c2cc182f66dac

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127346 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46994 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38129 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134410 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78901 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129218 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96922 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64920 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130294 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114062 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77416 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36952 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32178 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77792 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/119383 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136893 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/125809 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54005 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41612 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105440 "Found 1 new test failure: imported/blink/fast/pagination/viewport-y-horizontal-bt-rtl.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110411 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105119 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26826 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50639 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29090 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51581 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53942 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60029 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/158847 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53175 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39720 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56633 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54935 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
```diff
@@ -458,16 +458,16 @@ static void encodeObject(WKRemoteObjectEncoder *encoder, id object)
         [NSException raise:NSInvalidArgumentException format:@"%@ does not conform to NSSecureCoding", object];
 
-    if (class_isMetaClass(object_getClass(object)))
+    if (class_isMetaClass(RetainPtr { object_getClass(object) }.get()))
```
I know we need to do this to make the checker happy but adding retain/release here seems silly.

cdumez replied:
@rniwa Do we have an incoming static analysis fix for this? Should I suppress instead?

geoffreygaren replied:
Our convention elsewhere is that ObjC classes are considered immortal. As long as the return type of objc_getClass is accurate, we should be able to fix analysis to ignore this case. So I would suppress and file a bug for analysis.

rniwa replied:
I'm working on a fix for this so suppressing it sounds good.

darinadler commented on Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
```diff
@@ -176,9 +176,14 @@ - (NSString *)identifier
-    auto result = adoptNS([[NSMutableString alloc] initWithFormat:@"<%@: %p; protocol = \"%@\"; identifier = \"%@\"\n", NSStringFromClass(self.class), self, _identifier.get(), NSStringFromProtocol(_protocol)]);
+    auto result = adoptNS([[NSMutableString alloc] initWithFormat:@"<%@: %p; protocol = \"%@\"; identifier = \"%@\"\n", NSStringFromClass(self.class), self, _identifier.get(), NSStringFromProtocol(_protocol.get())]);
 
     auto descriptionForClasses = [](auto& allowedClasses) {
```
This should return a `RetainPtr` and not call `autorelease`.

Also wasteful that this creates and destroys `CString` objects just to compare two strings instead of calling `strcmp`.

cdumez replied:
Will address both issues.

darinadler commented on Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
```diff
@@ -64,15 +64,15 @@
-    address.get().country = nsStringNilIfNull(m_country);
-    address.get().ISOCountryCode = nsStringNilIfNull(m_isoCountryCode);
-    address.get().formattedAddress = nsStringNilIfNull(m_formattedAddress);
+    address.get().street = RetainPtr { nsStringNilIfNull(m_street) }.get();
```
Should `nsStringNilIfNull` return a `RetainPtr`?

cdumez replied:
Yes, it uses `autorelease()` internally anyway.

cdumez committed:
Address remaining safer CPP warnings in WebKit/Shared
https://bugs.webkit.org/show_bug.cgi?id=301163

Reviewed by Timothy Hatcher and Darin Adler.

* Source/WTF/wtf/text/WTFString.h:
(WTF::nsStringNilIfEmpty):
(WTF::nsStringNilIfNull):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper _accessibilityHitTestResolvingRemoteFrame:callback:]):
* Source/WebCore/platform/cocoa/DragImageCocoa.mm:
(WebCore::LinkImageLayout::LinkImageLayout):
* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::protectedPlatformPath const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::setVideoGravity):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerEnumsCocoa.mm:
(WebCore::convertMediaPlayerToAVLayerVideoGravity):
* Source/WebCore/platform/network/cf/ResourceError.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::parseDOMCookie):
(WebCore::NetworkStorageSession::setCookiesFromDOM const):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
* Source/WebCore/platform/network/mac/ResourceErrorMac.mm:
(WebCore::ResourceError::ResourceError):
(WebCore::ResourceError::mapPlatformError):
(WebCore::ResourceError::cfError const):
(WebCore::ResourceError::protectedCFError const):
* Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h:
(WebGPU_Internal::convertWTFStringToNSString):
* Source/WebKit/Platform/ios/PaymentAuthorizationController.mm:
(-[WKPaymentAuthorizationControllerDelegate presentationSceneIdentifierForPaymentAuthorizationController:]):
(-[WKPaymentAuthorizationControllerDelegate presentationSceneBundleIdentifierForPaymentAuthorizationController:]):
* Source/WebKit/SaferCPPExpectations/MemoryUnsafeCastCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/NoUnretainedMemberCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebKit/Shared/API/Cocoa/WKRemoteObjectCoder.mm:
(encodeObject):
* Source/WebKit/Shared/API/Cocoa/_WKRemoteObjectInterface.mm:
(-[_WKRemoteObjectInterface initWithProtocol:identifier:]):
(-[_WKRemoteObjectInterface protocol]):
(-[_WKRemoteObjectInterface debugDescription]):
* Source/WebKit/Shared/API/c/cf/WKStringCF.mm:
(WKStringCreateWithCFString):
(wkNSStringClassSingleton): Deleted.
* Source/WebKit/Shared/API/c/cf/WKURLCF.mm:
(WKURLCreateWithCFURL):
(wkNSURLClassSingleton): Deleted.
* Source/WebKit/Shared/API/c/mac/WKURLRequestNS.mm:
(WKURLRequestCopyNSURLRequest):
* Source/WebKit/Shared/API/c/mac/WKURLResponseNS.mm:
(WKURLResponseCopyNSURLResponse):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::typeFromObject):
* Source/WebKit/Shared/Cocoa/CoreIPCContacts.mm:
(WebKit::CoreIPCCNPostalAddress::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPassKit.mm:
(WebKit::CoreIPCPKContact::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCPersonNameComponents.mm:
(WebKit::CoreIPCPersonNameComponents::toID const):
* Source/WebKit/Shared/Cocoa/RevealItem.h:
* Source/WebKit/Shared/Cocoa/RevealItem.mm:
(WebKit::RevealItem::highlightRange const):
(WebKit::RevealItem::protectedItem const):
* Source/WebKit/Shared/Cocoa/WKNSError.mm:
(-[WKNSError _web_createTarget]):
* Source/WebKit/Shared/Cocoa/WKNSURLRequest.mm:
(-[WKNSURLRequest _web_createTarget]):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/Daemon/PCMDaemonEntryPoint.mm:
(WebKit::CompletionHandler<void):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::toCAFilterType):
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
(WebKit::RemoteLayerTreePropertyApplier::applyProperties):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::AuxiliaryProcess::launchServicesCheckIn):
(WebKit::webKit2BundleSingleton):
(WebKit::getSandboxProfileOrProfilePath):
(WebKit::populateSandboxInitializationParameters):
(WebKit::webKit2Bundle): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(-[WKWebExtension defaultLocale]):
(-[WKWebExtension displayName]):
(-[WKWebExtension displayShortName]):
(-[WKWebExtension displayVersion]):
(-[WKWebExtension displayDescription]):
(-[WKWebExtension displayActionLabel]):
(-[WKWebExtension version]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _getTextFragmentMatchWithCompletionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _applicationNameForDesktopUserAgent]):
(-[WKWebViewConfiguration applicationNameForUserAgent]):
(-[WKWebViewConfiguration _groupIdentifier]):
* Source/WebKit/UIProcess/API/Cocoa/_WKContextMenuElementInfo.mm:
(-[_WKContextMenuElementInfo qrCodePayloadString]):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _defaultSTSLabel]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(+[WKAutocorrectionContext autocorrectionContextWithWebContext:]):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
* Source/WebKitLegacy/mac/History/WebHistoryItem.mm:
(-[WebHistoryItem URLString]):
(-[WebHistoryItem originalURLString]):
(-[WebHistoryItem title]):
(-[WebHistoryItem alternateTitle]):
* Source/WebKitLegacy/mac/WebView/WebHTMLRepresentation.mm:
(-[WebHTMLRepresentation title]):
* Source/WebKitLegacy/mac/WebView/WebScriptDebugDelegate.mm:
(-[WebScriptCallFrame functionName]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _mainFrameOverrideEncoding]):

Canonical link: https://commits.webkit.org/301920@main
